### PR TITLE
fix DualMode ReceiveFrom on macOS

### LINKED ISSUE #26867: Sockets.ReceiveFrom should use recvfrom?
The `ReceiveFrom` implementation is sharing the same underling implementation as `ReceiveMessageFrom` even through `ReceiveFrom` doesn't care about the packet information... 

A side effect is that on macOS when using Dual mode sockets `ReceiveFrom` throws `System.PlatformNotSupportedException` because getting packet information doesn't work with dual mode sockets (see dotnet/runtime#15501).

It would be better and more efficient to implement `ReceiveFrom` using the platform specific primitive `recvfrom` rather than `recvmsg` which probably comes at a higher cost. This would also allow using `ReceiveFrom` with dual mode sockets on macOS.

### PR DESCRIPTION
**Pull request by wfurt**
fixes  #26867 
The ReceiveFrom works on macOs. The problem is that when the remote address is IPv4 on DualMode socket it comes back as plain old IPv4 so we fail to create the endpoint. It does come as IPv4Mapped Ipv6 address on Linux & Windows. 
My initial fix was just to handle it as IPv4 but for consistency I decided to turn it into mapped address as well. 
That should match `LocalEndpoint` and  `RemoteEndPoint` on `DualMode` `Socket` that also comes as mapped addresses.

### CODE REVIEW

File: src/libraries/Common/src/System/Net/SocketAddress.cs [hunk 1]
```diff
@@ -164,11 +164,14 @@ internal IPAddress GetIPAddress()
             }
         }
 
+        internal int GetPort()
+        {
+            return (int)SocketAddressPal.GetPort(Buffer);
+        }
```
Comment (liveans) [inline src/libraries/Common/src/System/Net/SocketAddress.cs:170]:
```suggestion
        internal int GetPort() => (int)SocketAddressPal.GetPort(Buffer);
```

File: src/libraries/Common/src/System/Net/SocketAddress.cs [hunk 2]
```diff
@@ -164,11 +164,14 @@ internal IPAddress GetIPAddress()
...
+
         internal IPEndPoint GetIPEndPoint()
         {
-            IPAddress address = GetIPAddress();
-            int port = (int)SocketAddressPal.GetPort(Buffer);
-            return new IPEndPoint(address, port);
+            return new IPEndPoint(GetIPAddress(), GetPort());
```
Comment (liveans) [inline src/libraries/Common/src/System/Net/SocketAddress.cs:174]: We can use expression body definition syntax here

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 3]
```diff
@@ -1687,8 +1687,6 @@ public int ReceiveFrom(byte[] buffer, int offset, int size, SocketFlags socketFl
             ValidateBufferArguments(buffer, offset, size);
             ValidateReceiveFromEndpointAndState(remoteEP, nameof(remoteEP));
 
-            SocketPal.CheckDualModeReceiveSupport(this);
```
Comment (antonfirsov) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1690]: Can we rename this to `CheckDualModePacketInfoSupport`?

Comment (wfurt) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:1690]: updated

File: src/libraries/Common/src/System/Net/SocketAddress.cs (hunk 2 above)
Comment (wfurt) [reply src/libraries/Common/src/System/Net/SocketAddress.cs:174]: updated.

**Review by liveans — APPROVED** [review]
Thanks!

**Review by antonfirsov — COMMENTED** [review]
There are some accidental changes, otherwise LGTM.

File: src/native/libs/System.Native/pal_networking.c [hunk 4]
```diff
@@ -1399,8 +1399,13 @@ int32_t SystemNative_ReceiveMessage(intptr_t socket, MessageHeader* messageHeade
     ConvertMessageHeaderToMsghdr(&header, messageHeader, fd);
 
     ssize_t res;
+    printf("%s:%d: recvmsg name %p %d socksize=%zd\n", __func__, __LINE__, header.msg_name, header.msg_namelen, sizeof(struct sockaddr_in6));
     while ((res = recvmsg(fd, &header, socketFlags)) < 0 && errno == EINTR);
 
+    printf("%s:%d: recvmsg finished with %zd name %p %d\n", __func__, __LINE__, res, header.msg_name, header.msg_namelen);
+    printf("%s:%d: poer=%du addr (%s)\n", __func__, __LINE__, ((struct sockaddr_in*)header.msg_name)->sin_port, inet_ntoa(((struct sockaddr_in*)header.msg_name)->sin_addr));
```
[the remark this answers is not in the document]
Comment (wfurt) [reply src/native/libs/System.Native/pal_networking.c]: Where do you see it @antonfirsov. This was removed yesterday in https://github.com/dotnet/runtime/pull/78261/commits/70a11fa8bb087e89932f1c62595dc13d5f433d18

[the remark this answers is not in the document]
Comment (antonfirsov) [reply src/native/libs/System.Native/pal_networking.c]: Not sure how I ended up commenting on the outdated variant.

**Review by antonfirsov — APPROVED** [review]

**Pull request merged**